test/foundry/PriceFeedDispatcher.t.sol

```diff
@@ -0,0 +1,203 @@
+        return new PriceFeedDispatcher(UniswapV3PriceFeed(0), _chainlinkPriceFeed);
+    }
+}
+
+contract PriceFeedDispatcherConstructorTest is PriceFeedDispatcherSetup {
+    function test_UniswapV3PriceFeed_can_be_zero_address() public {
+        _priceFeedDispatcher = _create_PriceFeedDispatcherUniswapV3PriceFeedNotExist();
+    }
+
+    function test_PFD_UECOU() public {
+        vm.expectRevert(bytes("PFD_UECOU"));
+        _priceFeedDispatcher = new PriceFeedDispatcher(UniswapV3PriceFeed(makeAddr("HA")), _chainlinkPriceFeed);
+    }
+
+    function test_PFD_CNC() public {
+        vm.expectRevert(bytes("PFD_CNC"));
+        _priceFeedDispatcher = new PriceFeedDispatcher(_uniswapV3PriceFeed, ChainlinkPriceFeedV3(0));
+    }
+}
+
+contract PriceFeedDispatcherTest is IPriceFeedDispatcherEvent, PriceFeedDispatcherSetup {
+    address public nonOwnerAddress = makeAddr("nonOwnerAddress");
+    uint256 internal _chainlinkPrice = 100 * 1e18;
+    uint256 internal _uniswapPrice = 50 * 1e18;
+
+    function setUp() public virtual override {
+        PriceFeedDispatcherSetup.setUp();
+        _priceFeedDispatcher = _create_PriceFeedDispatcher();
+        _priceFeedDispatcherUniswapV3PriceFeedNotExist = _create_PriceFeedDispatcherUniswapV3PriceFeedNotExist();
+
+        vm.mockCall(
+            address(_uniswapV3PriceFeed),
+            abi.encodeWithSelector(_uniswapV3PriceFeed.getPrice.selector),
+            abi.encode(50 * 1e8)
+        );
+
+        vm.mockCall(
+            address(_uniswapV3PriceFeed),
+            abi.encodeWithSelector(_uniswapV3PriceFeed.decimals.selector),
+            abi.encode(8)
+        );
+
+        vm.mockCall(
+            address(_chainlinkPriceFeed),
+            abi.encodeWithSelector(_chainlinkPriceFeed.getCachedTwap.selector),
+            abi.encode(100 * 1e8)
+        );
+    }
+
+    function test_setPriceFeedStatus() public {
+        _expect_emit_event_from_PriceFeedDispatcher();
+        emit StatusUpdated(Status.Chainlink);
+        _priceFeedDispatcher.setPriceFeedStatus(Status.Chainlink);
+        assertEq(uint256(_priceFeedDispatcher.getStatus()), uint256(Status.Chainlink));
+
+        emit StatusUpdated(Status.UniswapV3);
+        _priceFeedDispatcher.setPriceFeedStatus(Status.UniswapV3);
+        assertEq(uint256(_priceFeedDispatcher.getStatus()), uint256(Status.UniswapV3));
+    }
+
+    function test_cannot_setPriceFeedStatus_by_nonOwnerAddress() public {
+        vm.expectRevert(bytes("Ownable: caller is not the owner"));
+        vm.prank(nonOwnerAddress);
+        _priceFeedDispatcher.setPriceFeedStatus(Status.UniswapV3);
+    }
+
+    function test_cannot_setPriceFeedStatus_when__uniswapV3PriceFeed_not_exist() public {
+        vm.expectRevert(bytes("PFD_UU"));
+        _priceFeedDispatcherUniswapV3PriceFeedNotExist.setPriceFeedStatus(Status.UniswapV3);
+    }
+
+    function test_setChainlinkPriceFeedV3() public {
```

Comment:
In the begining of the spec, we don't have any ownerable feature. I forgot why we added ownerable after some discussion with @tienshaoku when implementing the setStatus() to turn the uniswap back to chainlink. For this setter and uniswap setter are just for the convenience. Because we have the ownerable to easily adjust the states in some rare situations. Otherwise, ownerable is a bit unnecessary. Maybe we should just use the initial spec to redploy when we get falled back to uniswap.

Reply:
Fixed in [7f52911](https://github.com/perpetual-protocol/perp-oracle-contract/pull/57/commits/7f529118a9c0e56fbb5e9671daebb7d535adaaa8)